### PR TITLE
allow arbitrary code from validator object

### PR DESCRIPTION
**Pull request by javierguerragiraldez**
the purpose is to enable library user to inject in-memory patches to the
Lua environment.

### CODE REVIEW

File: internal/vm/vm.go [hunk 1]
```diff
@@ -79,6 +80,22 @@ func (v *VM) Execute(file io.Reader, filename string) error {
...
+	obj := v.l.Get(lua.GlobalsIndex)
+	for _, n := range strings.Split(name, ".") {
+		if obj == lua.LNil {
+			break
+		}
+		obj = v.l.GetField(obj, n)
+	}
+	return obj
+}
+
```
Comment (mikefero) [inline internal/vm/vm.go:98]: I know you covered a `panic` in a unit test, but would it be better if we handle the panic in this function and return an error if it occurred? My reasoning here is that we can act upon the error more gracefully than the panic.

```suggestion
// deepGet("a.b.c") is equivalent to lua's "a.b.c".
// More exactly, like "a and a.b and a.b.c".
// Note that if 'a' or 'a.b' isn't a table (or
// anything with a metatable), it still raises
// an error (just like in Lua).
func (v *VM) deepGet(name string) (lua.LValue, error) {
	defer func() {
		if err := recover(); err != nil {
			return lua.LNil, err
		}
	}()

	obj := v.l.Get(lua.GlobalsIndex)
	for _, n := range strings.Split(name, ".") {
		if obj == lua.LNil {
			break
		}
		obj = v.l.GetField(obj, n)
	}
	return obj, nil
}

```

**Review by mikefero — COMMENTED** [review]
This will be a nice feature. One question/suggestion; I might be looking at it wrong though.

File: internal/vm/vm.go (hunk 1 above)
Comment (javierguerragiraldez) [reply internal/vm/vm.go:98]: i did consider this, but decided to try a bit to not panic at all in the first place.  in the other test, i actually do it by replacing the Panic callback from the VM with a dummy function.  of course, PCall must do something similar but more "complete", perhaps i could find some primitive to use without having to wrap in a Lua function...

but no.  `lua.PCall()` doesn't use any primitive, it does it "directly"... and it's a monstrosity that not only does panic/recover, but also a lot of stack meddling to pass the error back to the Lua state (or worse, call a lua error handler).

better to use your solution

Comment (javierguerragiraldez) [reply internal/vm/vm.go]: ... and of course, a `defer` function can't just return stuff.  have to use labelled returns for that

**Comment by hbagdi** [comment]
@javierguerragiraldez @mikefero A ping to get this going.

File: internal/vm/vm.go (hunk 1 above)
Comment (mikefero) [reply internal/vm/vm.go]: > have to use labelled returns for that

whoops, yeah my code can't be trusted sometimes ;)

**Comment by mikefero** [comment]
@javierguerragiraldez could we clean up the commit message:

- Commit 1
```
feat: allow arbitrary Lua code execution

This feature will enable library users to execute Lua code and perform
in-memory patches to the Lua environment.
```
Something like that

- Commit 2 can be squashed with commit 1
- Commit 3
```
chore: bump up go to 1.19 and golangci-lint to 1.48

Notable changes:
- Spelling corrections
- ioutil package is deprecated and is not used anymore
```
Something like that

**Review by mikefero — APPROVED** [review]
Looks good, my only nit is

```
feat: allow arbitrary Lua code from validator object
```

`goks` doesn't have the validator object which is why I was going for something like the following for the heading of the commit:

```
feat: allow arbitrary Lua code execution
```

**Comment by javierguerragiraldez** [comment]
> `goks` doesn't have the validator object

this is in goks/plugin/validator.go file, adding a new `func (v *Validator) Execute(....` method.   Or do you prefer "Validator struct" instead?

@hbagdi the commits are separated between "new feature" and "fix linter"

**Comment by hbagdi** [comment]
> this is in goks/plugin/validator.go file, adding a new func (v *Validator) Execute(.... method. Or do you prefer "Validator struct" instead?

Indeed. We are all good.

**Pull request merged**